src/gpuarray_array.c

```diff
@@ -1008,15 +1009,18 @@ int GpuArray_concatenate(GpuArray *r, const GpuArray **as, size_t n,
 
   res_off = r->offset;
   res_dims = r->dimensions;
+  res_flags = r->flags;
+  r->flags &= ~(GA_C_CONTIGUOUS|GA_F_CONTIGUOUS);
```

Comment:
1) Once we decide that we need to do collapsing it's too late and the code is already on the path for non-contiguous.
2) Also the collapsing code has currently no way to indicate that what we work on is actually contiguous.
3) Finally the purpose of those flags is to avoid having to do checks like what you propose.

For these three reason I still oppose adding a dirty hack for an obscure corner case.